src/Market/MarketHandler.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+declare(strict_types=1);
+
+namespace EtoA\Market;
+
+use EtoA\Core\Configuration\ConfigurationService;
+use EtoA\Fleet\FleetRepository;
+use EtoA\Support\RuntimeDataStore;
+use EtoA\Universe\Planet\PlanetRepository;
+use EtoA\Universe\Resources\BaseResources;
+
+class MarketHandler
+{
+    private MarketRateRepository $marketRateRepository;
+    private RuntimeDataStore $runtimeDataStore;
+    private PlanetRepository $planetRepository;
+    private FleetRepository $fleetRepository;
+    private ConfigurationService $config;
+
+    public function __construct(
+        MarketRateRepository $marketRateRepository,
+        RuntimeDataStore $runtimeDataStore,
+        PlanetRepository $planetRepository,
+        FleetRepository $fleetRepository,
+        ConfigurationService $config
+    ) {
+        $this->marketRateRepository = $marketRateRepository;
+        $this->runtimeDataStore = $runtimeDataStore;
+        $this->planetRepository = $planetRepository;
+        $this->fleetRepository = $fleetRepository;
+        $this->config = $config;
+    }
+
+    /**
+     * @return float[]
+     */
+    public function calcRate(): array
+    {
+        $rateMin = $this->config->getFloat('market_rate_min');
+        $rateMax = $this->config->getFloat('market_rate_max');
+
+        // Resulting rates
+        $rates = array_fill(0, BaseResources::NUM_RESOURCES, 1);
+        $res_rates = array_fill(0, BaseResources::NUM_RESOURCES, 1);
+
+        // Load previous rates
+        $previousRates = $this->marketRateRepository->getRates($this->config->getInt('market_rates_count'));
+        $nr = count($previousRates);
+        $rc = 0;
+        foreach ($previousRates as $rate) {
+            // Weight factor calculation. Insert any other formula if you desire so.
+            // Newer values should weight more than older values
+            $factor = $nr - $rc;
+            //$factor = log($nr-$rc);
+
+            // For every resource calculate the ration and multiply with the weight factor
+            for ($i = 0; $i < BaseResources::NUM_RESOURCES; $i++) {
+                $supplyProp = "supply$i";
+                $demandProp = "demand$i";
+                if ($rate->$supplyProp > 0) {
+                    $r = ($rate->$demandProp / $rate->$supplyProp);
```

Comment:
Not quite sure how to best avoid this but I think in the long run we should remove dynamic code like this

Reply:
Yes agree with you, will require db schema changes which i try to avoid for the moment